### ISSUE TITLE
Implement the Data Pruning at the file level of Data Lake formats

### ISSUE THREAD

**Issue by alifirat** [ClickHouse/ClickHouse#52054]
**Use case**

Today, ClickHouse supports the read from the Data Lake formats but do not support the pruning of the data files to read. 

**Describe the solution you'd like**

Add partition keys to the Data Lake tables that will be used for the data pruning of the files. 

**Describe alternatives you've considered**

Either add a new component on top of ClickHouse that does the data pruning, Either use a different Compute Engine like Trino that supports the Data pruning with Iceberg for instance. 

**Additional context**

To prove that no data pruning is applied on the files to read, I can share you a small request that was made on a table partionned by day. On a 3 days of data, this is the information from the system.query_log
```
duration_secs: 28.786
read_rows: 457427120
memory_usage: 165.53 MiB
read_bytes: 113.70 GiB
read_from_s3: 17.14 GiB
s3_read_requests_count: 6077
s3_head_object: 276
s3_get_object: 5792
```

On a single day:

```
duration_secs: 30.128
read_rows: 457427120
memory_usage: 166.08 MiB
read_bytes: 113.70 GiB
read_from_s3: 17.14 GiB
s3_read_requests_count: 6077
s3_head_object: 276
s3_get_object: 5792
```

**Comment by alanpaulkwan** [comment]
AFAIK its just reading the Parquet files directly, and tests show it's about the same speed.

**Comment by alifirat** [comment]
Yep and the idea is to read less Parquet files when the Iceberg table is partitioned.

**Comment by alifirat** [comment]
Hey @Avogar 

I was using a dataset generated by spark that was partitionned by data then I've created my table in ClickHouse using the `Iceberg` Engine. At this step, I didn't specify any partitions because the Iceberg Engine doesn't support it and anyway the partitions informations are in the metadata manifest. 

I didn'y any performance improvement between the following queries: 
```select sum(sign) from my_iceberg_test where dt = yesterday() - 7 and id = 10```  and 
```select sum(sign) from my_iceberg_test where dt = yesterday() - 3 and id = 10```

To me ClickHouse has not implemented the data pruning on the file level.

**Comment by Avogar** [comment]
Yes, Iceberg specific data pruning on the file level is not implemented. But AFAIK Iceberg files contains partition in the file path, so you can filter it by filtering virtual column `_path`.
So, if your data is partitioned by `dt` in days, there should be folder with name `.../dt=x/...`, so we can use smth like:
```
select sum(sign) from my_iceberg_test where _path like concat('%dt=', toString(yesterday() - 7), '%') and id = 10
```
And ClickHouse will filter out those paths. I know that it's not natural and requires some additional knowledge. But the problem is that Implementing real Iceberg data pruning is not so easy and it requires some effort, and we still not sure if it's worth it.

**Comment by alifirat** [comment]
Multiple points regarding your comment: 
1. Can you please it make it clear, in the docs, that there is at least one way to use the partition feature with Iceberg tables  using the `virtual_path`. At least, the future users will be aware of that. 
2. Is it bit to bad to have implement the feature of supporting Iceberg partially because usually the companies comes with a ton of data and the partitions in Iceberg, like in ClickHouse, is a key feature. To me it's opening new doors for ClickHouse to be used as ComputeEngine with the data lakes formats.

**Comment by TechTinkerer42** [comment]
Apart from containing partition information, Iceberg file metadata includes additional metrics such as lower and upper bounds, as well as counts of null, NaN, and total values.
Consequently, the workaround using a virtual column _path is ineffective for these metrics.

https://tabular.io/blog/iceberg-metadata-indexing/
> Iceberg stores column-level metrics for each data file for an extra level of filtering within each partition. The metrics kept are lower and upper bounds, as well as null, NaN, and total counts for the columns in a file. After filtering by partition, Iceberg will check whether each file matches a given query filter using these stats. File formats will also use column stats to skip rows

**Comment by Avogar** [comment]
We have plans to improve Iceberg integration this year.

**Comment by Blackmorse** [comment]
> We have plans to improve Iceberg integration this year.

Fingers crossed :)

**Comment by jgprogramming** [comment]
@Avogar  any updates on this issue? We really would like to use clickhouse both as data storage and a query engine for cold data but lack of iceberg optimizations is preventing that usecase.

**Comment by Avogar** [comment]
@divanik is going to start working on Iceberg improvements including data pruning soon.

**Comment by melvynator** [comment]
This feature is planed for Q4.

**Comment by alexey-milovidov** [comment]
Implemented in #72044, available in 25.1.

**Comment by divanik** [comment]
It is not completed completely for Iceberg, only for time datatypes

**Comment by divanik** [comment]
Here is is: https://github.com/ClickHouse/ClickHouse/issues/75060
